src/blockchain/rpc.rs

```diff
@@ -495,6 +285,507 @@ fn list_wallet_dir(client: &Client) -> Result<Vec<String>, Error> {
+                        outpoint: OutPoint::new(tx_res.info.txid, tx_res.detail.vout),
+                        txout,
+                        keychain,
+                        is_spent: false,
+                    };
+                    self.updated_utxos.insert(utxo);
+                    self.update_last_index(keychain, index);
+                }
+            }
+
+            // update tx deltas
+            self.retained_txs.insert(tx_res.info.txid);
+            if updated {
+                self.updated_txs.insert(tx_res.info.txid);
+            }
+        }
+
+        // update tx sent fields from tx inputs
+        self.txs
+            .clone() // clone is required as we cannot have more than two mutable references
+            .values()
+            .try_for_each(|db_tx| -> Result<(), Error> {
+                // only bother to update txs that are retained
+                if self.retained_txs.get(&db_tx.txid).is_some() {
+                    // if we do not have the raw_tx, we cannot calculate the sent value
+                    if let Some(raw_tx) = db_tx.transaction.as_ref() {
+                        let sent = self.sent_from_raw_tx(raw_tx)?;
+                        self.txs
+                            .entry(db_tx.txid)
+                            .and_modify(|db_tx| db_tx.sent = sent);
+                    }
+                }
+                Ok(())
+            })?;
+
+        // obtain UTXOs from Core wallet
+        let core_utxos = client
+            .list_unspent(Some(0), None, None, Some(true), None)?
+            .into_iter()
+            .filter_map(|utxo_entry| {
+                let path_result = self
+                    .db
+                    .get_path_from_script_pubkey(&utxo_entry.script_pub_key)
+                    .transpose()?;
+
+                let utxo_result = match path_result {
+                    Ok((keychain, index)) => {
+                        self.update_last_index(keychain, index);
+                        Ok(Self::make_local_utxo(utxo_entry, keychain, false))
+                    }
+                    Err(err) => Err(err),
+                };
+
+                Some(utxo_result)
+            })
+            .collect::<Result<HashSet<_>, Error>>()?;
+
+        // mark "spent utxos" to be updated in database
+        let spent_utxos = self.utxos.difference(&core_utxos).cloned().map(|mut utxo| {
+            utxo.is_spent = true;
+            utxo
+        });
+
+        // mark new utxos to be added in database
+        let new_utxos = core_utxos.difference(&self.utxos).cloned();
+
+        // add to updated utxos
+        self.updated_utxos.extend(spent_utxos.chain(new_utxos));
+
+        Ok(self)
+    }
+
+    /// Calculates received amount from raw tx.
+    fn received_from_raw_tx(db: &D, raw_tx: &Transaction) -> Result<u64, Error> {
+        raw_tx.output.iter().try_fold(0_u64, |recv, txo| {
+            let v = if db.is_mine(&txo.script_pubkey)? {
```

Comment:
Looking at how `received_from_raw_tx` is used, it is used while a mutable reference to a field of `self` still exists.

The suggested change will result in:
```
error[E0502]: cannot borrow `*self` as immutable because it is also borrowed as mutable
```